### ISSUE TITLE
RFC: Use Dovecot director proxy

### ISSUE THREAD

**Issue by muhlemmer** [Mailu/Mailu#1222]
Currently, we are using Nginx as the sole reverse proxy for TLS and IMAP/SMTP/HTTP authentication termination. Nginx is capable of talking IMAP and SMTP protocols.

The current has a known limitation: Nginx is not capable to proxy manage-sieve. A stagnated PR is already open for some time: #428.

Furthermore, I'm investigating replicated setups for Dovecot. When using a central Maildir storage for concurrent access, Dovecot recommends the use of a [Director service](https://wiki2.dovecot.org/Director). This makes sure that a user session is always directed to the same Dovecot back-end.

In replicated environments the Dovecot director proxy will need to be host bound (so not Docker `ingress` network mash), in order to guarantee persistence from client till director. This makes it also impossible to use Nginx as TLS over TCP proxy.

The proposed design will have the following challenges:
- Certificates need to be shared between Nginx and Dovecot
- Dovecot director needs to reload on certificate writes
- Dovecot director needs to be integrated with podop for authentication calls

@kaiyou: As I understand from previous discussions, you are not the biggest fan of this. But I do think that this approach will give us more freedom to scale Mailu and implement manage-sieve in a secure way. I hope you reconsider ;).

**Comment by kaiyou** [comment]
Indeed I am not the biggest fan :smile: 
I could rethink it though, but I really insist on us having a somehow unified frontend that handles authentication and reverse proxying.
Maybe I could come around and admit that splitting the Web and mail realms is doable. It would require that Dovecot director interacts with podop indeed and has access to certificates.
Random ideas that we need to explore in that direction:
 - getting back to traefik now that traefik 2 supports TCP reverse proxying with TLS (I have not yet tried combining TCP TLS and letsencrypt though)
 - using Dovecot as a single frontend for email, including the MSA reverse proxy instead of nginx (https://doc.dovecot.org/admin_manual/submission_server/)

I will keep posting ideas, but this needs a lot of prototyping :)

**Comment by micw** [comment]
I don't think that trafik 2 can replace nginx fully since besides SSL we also have STARTTLS which needs to be supported at application level while SSL is at transport level.

Personally I think it would be a good idea to create an independent branch of mailu with all stuff needed for replicated setup so that it can be fullly tested.

**Comment by kaiyou** [comment]
Ah. Yep, forgot about starttls, which is why we dropped traefik in the first place, sorry.

**Comment by muhlemmer** [comment]
> But I really fear about the certificate hell then.

I think this is where we will enter in the conflicts of different deployment targets. On `docker-compose` systems the most straight forward this is using `nginx` + letsencrypt. However, as I briefly explained [here](https://usrpro.com/publications/mailu-persistent-storage/#current-storage-situation-2), in any replicated situation we can't have multiple nginx instances requesting certificates. Because:

1. It will be a race condition on the certificate files
2. The ACME challenge is kept local. If the ACME server calls back for the challenge, there is a probablity that it connects to the wrong instance and it will not find the correct challenge.

(Traefik deals with this problem using a centralized key/value store, but my personal experience isn't too good with that)

In replicated situations the admin will have to find alternative means of getting certificates and sending them to `front`. After which it will need to be reloaded. Currently we have a listener script inside nginx for exactly that. This might also be of good use for Dovecot / whatever solution we might pick.

**Comment by micw** [comment]
Traefik's "clustering" caused lots of cert issue failures in our instalaltions. We finally switched to certmanager which works very well on a replicated environment (and works fine together with traefik)

**Comment by muhlemmer** [comment]
> Traefik's "clustering" caused lots of cert issue failures in our installations. We finally switched to certmanager which works very well on a replicated environment (and works fine together with traefik)

Yep, same here. That's what I meant by "not to good" :sweat:

**Comment by Nebukadneza** [comment]
Hi There,

The `Mailu`-Project is currently in a bit of a bind! We are short on man-power, and we need to judge if it is possible for us to put in some work on this issue.

To help with that, we are currently trying to find out which issues are actively keeping users from using `Mailu`, which issues have someone who want to work on them — and which issues may be less important. These a less important ones could be discarded for the time being, until the project is in a more stable and regular state once again.

In order for us to better assess this, it would be helpful if you could put a **reaction on this post** (use the :smiley: icon to the top-right).
- 👍️ if you **need this** to be able to use Mailu. Ideally, you’d also be able to test this on your installation, and provide feedback …
- 🎉 if you find it a **nice bonus**, but no deal-breaker
- 🚀 if you want to **work on it yourself**!
We want to keep this **voting open for 2 weeks** from now, so please help out!

**Comment by nextgens** [comment]
A basic PoC has been implemented in #2790 for sieve... there is no reason why we couldn't do the same to replace ngx_mail elsewhere.

Technically we don't use the Director plugin (just the proxy) but admin is in control of where things get redirected... so we could trivially implement director-like features there if we wanted to... or use the real thing.

**Comment by nextgens** [comment]
Master now runs a dovecot proxy for everything. I have given hints on what needs to happen to get "director like" functionality.

For the record, the next version of dovecot won't have a Director any-more (hence why deploying theirs is not an option).

https://github.com/Mailu/helm-charts/pull/304#issuecomment-1779618667
